tensorboard/plugins/debugger_v2/tf_debugger_v2_plugin/store/debugger_reducers.ts

```diff
@@ -674,6 +675,102 @@ const reducer = createReducer(
       };
     }
   ),
+  ////////////////////////////////////////////////
+  // Reducers related to graph structures.      //
+  ////////////////////////////////////////////////
+  on(
+    actions.graphOpFocused,
+    (
+      state: DebuggerState,
+      data: {graph_id: string; op_name: string}
+    ): DebuggerState => {
+      return {
+        ...state,
+        graphs: {
+          ...state.graphs,
+          focusedOp: {
+            graphId: data.graph_id,
+            opName: data.op_name,
+          },
+        },
+      };
+    }
+  ),
+  on(
+    actions.graphOpInfoRequested,
+    (
+      state: DebuggerState,
+      data: {graph_id: string; op_name: string}
+    ): DebuggerState => {
+      const {graph_id, op_name} = data;
+      const newState: DebuggerState = {
+        ...state,
+        graphs: {
+          ...state.graphs,
+          loadingOps: {
+            ...state.graphs.loadingOps,
+          },
+        },
+      };
+      if (newState.graphs.loadingOps[graph_id] === undefined) {
+        newState.graphs.loadingOps[graph_id] = {};
+      }
+      if (newState.graphs.loadingOps[graph_id][op_name] === undefined) {
+        newState.graphs.loadingOps[graph_id][op_name] = DataLoadState.LOADING;
+      }
+      return newState;
+    }
+  ),
+  on(
+    actions.graphOpInfoLoaded,
+    (state: DebuggerState, data): DebuggerState => {
+      const opInfo: GraphOpInfo = data.graphOpInfoResponse;
+      const {inputs, consumers, graph_ids} = opInfo;
+      const graphId = graph_ids[graph_ids.length - 1];
+      const newState: DebuggerState = {
+        ...state,
+        graphs: {
+          ...state.graphs,
+          ops: {
+            ...state.graphs.ops,
+          },
+          loadingOps: {
+            ...state.graphs.loadingOps,
+          },
+        },
+      };
+      for (const input of inputs) {
+        if (!input.data) {
+          continue;
+        }
+        if (newState.graphs.ops[graphId] === undefined) {
+          newState.graphs.ops[graphId] = {};
+        }
+        newState.graphs.ops[graphId][input.op_name] = input.data;
+        delete input.data;
+      }
+      for (let i = 0; i < consumers.length; ++i) {
+        for (const consumer of consumers[i]) {
+          if (!consumer.data) {
+            continue;
+          }
+          if (newState.graphs.ops[graphId] === undefined) {
+            newState.graphs.ops[graphId] = {};
+          }
+          newState.graphs.ops[graphId][consumer.op_name] = consumer.data;
+          delete consumer.data;
+        }
+      }
+      if (newState.graphs.ops[graphId] === undefined) {
+        newState.graphs.ops[graphId] = {};
+      }
+      newState.graphs.ops[graphId][opInfo.op_name] = opInfo;
+      // Remove the loading marker for the op.
+      newState.graphs.loadingOps[graphId][opInfo.op_name] =
```

Comment:
Logically impossible for this to be true but the check `if (newState.graphs.loadingOps[graph_id] === undefined) {` might lower cognitive overhead (I would not have to reason harder on other flows)

Reply:
I addressed this by using a similar approach as above:

```
        graphs: {
          ...state.graphs,
          ops: {
            ...state.graphs.ops,
            [graphId]: {
              ...state.graphs.ops[graphId],
            }
          },
          loadingOps: {
            ...state.graphs.loadingOps,
            [graphId]: {
              ...state.graphs.loadingOps[graphId],   // <--
            }
          },
        },
```